### ISSUE TITLE
Tensorflow_ranking with keras

### ISSUE THREAD

**Issue by korosig** [tensorflow/ranking#197]
Hi there,

I want to try Tensorflow_ranking with Keras. (https://github.com/tensorflow/ranking/tree/master/tensorflow_ranking/examples/keras) 

As a first step I have made a copy/paste from your Github repo, but I didn't find the TEST/TRAIN/VAL dataset. I tried to use this: tensorflow_ranking/examples/data but it does not work.

My real goal to make a model that is able to rank Students in different classes based on their achievements. (Numeric and categorical features)

My questions are:
- where could I find data to test code from tensorflow_ranking/examples/keras repo?
- how can I use tensorflow_ranking (with Keras) when the dataset is grouped? (Class by Class)

I want to test your code with Python IDE (Spyder) so I didn't install Basel and other things.

**Comment by azagniotov** [comment]
@korosig I think that data you want is under: https://github.com/tensorflow/ranking/tree/master/tensorflow_ranking/examples/data

**Comment by ramakumar1729** [comment]
@azagniotov Thanks for the accurate pointer to the data. 

@korosig : For data format, we use `ExampleListWithContext` proto as data format stored as TFRecords. See [demo](https://github.com/tensorflow/ranking/blob/master/tensorflow_ranking/examples/keras/keras_m2e_tfrecord.py) on how to train using this data.

**Comment by korosig** [comment]
> @korosig I think that data you want is under: https://github.com/tensorflow/ranking/tree/master/tensorflow_ranking/examples/data
> 
> when you say it does not work, what errors do you see?

I have solved the problem, but I don't understand how can I made a prediction with the trained model....

**Comment by azagniotov** [comment]
@korosig check the [sparse features tutorial](https://colab.research.google.com/github/tensorflow/ranking/blob/master/tensorflow_ranking/examples/handling_sparse_features.ipynb), `Generating Predictions` section. It shows how to make a prediction using your `estimator`, although the object is called `ranker` there.

Alternatively, you can deploy your model to a TensorFlow model server (https://www.tensorflow.org/tfx/serving/serving_basic) and make predictions over HTTP or gRPC

**Comment by korosig** [comment]
> @korosig I think that data you want is under: https://github.com/tensorflow/ranking/tree/master/tensorflow_ranking/examples/data
> 
> when you say it does not work, what errors do you see?

When I use _eval_input_fn()_ I got this:


({'document_tokens': <tensorflow.python.framework.sparse_tensor.SparseTensor at 0x23441080978>,
  'query_tokens': <tensorflow.python.framework.sparse_tensor.SparseTensor at 0x23441080b00>,
  'example_list_size': <tf.Tensor: shape=(10,), dtype=int32, numpy=array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2])>},
 <tf.Tensor: shape=(10, 2), dtype=float32, numpy=
 array([[1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.]], dtype=float32)>)

**Comment by korosig** [comment]
> @korosig check the [sparse features tutorial](https://colab.research.google.com/github/tensorflow/ranking/blob/master/tensorflow_ranking/examples/handling_sparse_features.ipynb), `Generating Predictions` section. It shows how to make a prediction using your `estimator`, although the object is called `ranker` there.
> 
> Or, you could read your saved model and make a request to it using a serialized ELWC (ExampleListWithContext) object:
> 
> ```
> saved_model_path = "<MODEL_NAME>/export/saved_model_exporter/<TIMESTAMP>"
> saved_model = tf.saved_model.load(saved_model_path)
> 
> # you can also pass your ELWC as a named parameter based on your 'receiver_name' in
> # 'build_ranking_serving_input_receiver_fn' if you have configured one. I think 'examples'
> # is the default name: examples=tf.constant([ELWC.SerializeToString()])
> output_dict_with_tensor = saved_model.signatures["predict"](tf.constant([ELWC.SerializeToString()]))
> 
> print("Predictions: ", output_dict_with_tensor['output'].numpy())
> ```
> 
> Alternatively, you can deploy your model to a TensorFlow model server (https://www.tensorflow.org/tfx/serving/serving_basic) and make predictions over HTTP or gRPC

If I change the _GROUP_SIZE = 1 to _GROUP_SIZE = 20 it means I train the model for LIST_WISE problem?

**Comment by korosig** [comment]
> @azagniotov Thanks for the accurate pointer to the data.
> 
> @korosig : For data format, we use `ExampleListWithContext` proto as data format stored as TFRecords. See [demo](https://github.com/tensorflow/ranking/blob/master/tensorflow_ranking/examples/keras/keras_m2e_tfrecord.py) on how to train using this data.

I have read every tutorial, look videos, and all stuff but I have no idea how can I transform my feature set into an ELWC data form. There is no description of this structure.

For example, a students profile is:
Student ID  age  grade  math%   physics%   english%  art%  math_competit  language_competit 
  14588        16     k12        98            67           88         100        first_place            very_good

If I have 20 students in the same class, how can I transform this data to be able to make a listwise prediction for every grade ( theoretically in every grade has 3 class with 20 students)

**Comment by azagniotov** [comment]
HI @korosig , 

In your feature set for 20 students, you need to decide what should be the context feature(s) and what should be example features 

That's the ELWC proto: https://github.com/tensorflow/serving/blob/master/tensorflow_serving/apis/input.proto#L19 Also, in the sparse features tutorial I linked earlier, there is a printout of ELWC, so you can see its structure 

In addition, have a look at https://github.com/tensorflow/ranking/issues/196 , perhaps this can give you ideas how to construct an ELWC.

**Comment by korosig** [comment]
> HI @korosig ,
> 
> In your feature set for 20 students, you need to decide what should be the context feature(s) and what should be example features. i.e.: you have 20 documents. So you need to decide what features belong to the query level (that's the context) and what features belong to the individual documents (that's your students)
> 
> That's the ELWC proto structure: https://github.com/tensorflow/serving/blob/master/tensorflow_serving/apis/input.proto#L19 Also, in the sparse features tutorial I linked earlier, there is a printout of ELWC, so you can see its structure
> 
> In addition, have a look at #196 , perhaps this can give you ideas how to construct an ELWC.

Thank you for your help :-)

if I understand correctly, the good solution could be to do a train / test split , where the train will be the Contex and the test will be Example, and each Train/Test will contain 20/20 students? In a 3D data structure?

**Comment by azagniotov** [comment]
@korosig In the ELWC format, the Context is inseparable from its Examples. That's the structure that allows you to describe a query (sorry, I keep using 'query' for my explanation) and its results, which are the documents (Examples).

Just a rough e.g.:
You said that a class can have 20 students. So, the class can be a feature in the Context (since the class is common across all the 20 students) and each student can be an Example under the Context.

**Comment by korosig** [comment]
> HI @korosig ,
> 
> In your feature set for 20 students, you need to decide what should be the context feature(s) and what should be example features
> 
> That's the ELWC proto: https://github.com/tensorflow/serving/blob/master/tensorflow_serving/apis/input.proto#L19 Also, in the sparse features tutorial I linked earlier, there is a printout of ELWC, so you can see its structure
> 
> In addition, have a look at #196 , perhaps this can give you ideas how to construct an ELWC.

In this case, how do I solve the fact that a student's result in a class is only enough for 5th place, until then he will be first elsewhere with the same result? Should I give an example of every possible option?

> @korosig In the ELWC format, the Context is inseparable from its Examples. That's the structure that allows you to describe a query (sorry, I keep using 'query' for my explanation) and its results, which are the documents (Examples).
> 
> Just a rough e.g.:
> You said that a class can have 20 students. So, the class can be a feature in the Context (since the class is common across all the 20 students) and each student can be an Example under the Context.

In this case, how do I solve the fact that a student's result in a class is only enough for 5th place, until then he will be first elsewhere with the same result? Should I give an example of every possible option?

**Comment by korosig** [comment]
> inseparable

The structure of ELWC is confused me. It is really not clear for me, and I didn't find any documentation about this. I do not have a real idea of how to transform a CSV (feature profile) into a ELWC data structure. Do you have any suggestions about this?

**Comment by azagniotov** [comment]
@korosig what are you trying to predict: a student's place within the class or the grade based on the student's results?

**Comment by korosig** [comment]
> @korosig what are you trying to predict: a student's place within the class or the grade based on the student's results?

Yes

**Comment by korosig** [comment]
[students.txt](https://github.com/tensorflow/ranking/files/4678841/students.txt)
 for example students rank

**Comment by azagniotov** [comment]
@korosig , I may be totally off here, so `pls take my suggestion with a pinch of salt`: 
It looks like you want to predict the `rank` within the grade, so `rank` can be your label. The `grade` can be your Context, while students with their grades within the class can be the Examples

```
examples {
  features {
    feature {
      key: "student_id"
      value {
        int64_list {
          value: 123
        }
      }
    }
    feature {
      key: "age"
      value {
        int64_list {
          value: 16
        }
      }
    }
    feature {
      key: "rank"
      value {
        int64_list {
          value: 5
        }
      }
    }
    feature {
      key: "geography"
      value {
        float_list {
          value: 91.96
        }
      }
    }
    feature {
      key: "art"
      value {
        float_list {
          value: 82.86
        }
      }
    }
    feature {
      key: "english"
      value {
        float_list {
          value: 67.10
        }
      }
    }
    feature {
      key: "physics"
      value {
        float_list {
          value: 82.60
        }
      }
    }
    feature {
      key: "math"
      value {
        float_list {
          value: 12.0
        }
      }
    }
  }
}
context {
  features {
    feature {
      key: "grade_name"
      value {
        bytes_list {
          value: "k12"
        }
      }
    }
  }
}
```

**Comment by charleystran** [comment]
@azagniotov I have a similar problem set and this makes alot of sense for me. What is still confusing from my end is that this represents 1 class. If there are 10 different classes in the problem set. How are do they putt together when training? Since the example is using tfrecord  I am having a hard time visualizing how that data is created and what is all included in it.

**Comment by azagniotov** [comment]
Hi @charleystran ,

You can try making an ELWC which has the class as a Context and each student as an Example. 

Then, let's say you have a five years worth of student data for those ten classes, so end up making a lot of ELWCs. You may end up with multiple ELWCs for the same class from different years

A single TF record can contain multiple ELWCs with different state; and not all ELWCs must be from the same class in the same TF record. Then you make three TF records from all your ELWCs for training, validation & testing.

**Comment by charleystran** [comment]
@azagniotov could you point me a good tutorial or documentation that adds multiple elwc’s to a tfrecord. I have been searching for a couple days and coming up blank.

**Comment by azagniotov** [comment]
@charleystran you may find this closed issue https://github.com/tensorflow/ranking/issues/196 useful. I provided there a code snippet that should point you to the right direction

**Comment by ramakumar1729** [comment]
Thanks for the detailed response, @azagniotov ! 

Closing this issue for now. Please feel free to reopen or create a new one.
